openstf/model/xgb_quantile.py

```diff
@@ -0,0 +1,102 @@
+#
+# SPDX-License-Identifier: MPL-2.0
+from typing import Tuple
+from functools import partial
+
+from sklearn.base import RegressorMixin, BaseEstimator
+from sklearn.utils.validation import check_X_y, check_array, check_is_fitted
+from xgboost import XGBRegressor
+import numpy as np
+
+import openstf.metrics.metrics as metrics
+
+DEFAULT_QUANTILES: Tuple[float, ...] = (0.9, 0.5, 0.1)
+
+
+class XgbQuantile(BaseEstimator, RegressorMixin):
+    def __init__(self, quantiles: Tuple[float, ...] = DEFAULT_QUANTILES):
+
+        # Check if quantile 0.5 is pressent this is required
+        if 0.5 not in quantiles:
+            raise ValueError(
+                "Cannot train quantile model as 0.5 is not in requested quantiles!"
+            )
+        else:
+            self.quantiles = quantiles
+
+    def fit(self, x: np.array, y: np.array) -> RegressorMixin:
+        """Fits xgb quantile model
+
+        Args:
+            x np.array: Feature matrix
+            y np.array: Labels
+
+        Returns:
+            Fitted XGBQuantile model
+
+        """
+
+        # Check/validate input
+        check_X_y(x, y)
+
+        # Convert input data to np.array (most of the time this is allready the case)
+        x = np.array(x)
+        y = np.array(y)
+
+        # Get fitting parameters
+        params_quantile = self.get_params().copy()
+
+        quantile_models = {}
+
+        for quantile in self.quantiles:
+            # Define objective callback functions specifically for desired quantile
+            params_quantile["obj"] = partial(
+                metrics.xgb_quantile_eval, quantile=quantile
+            )
+            params_quantile["feval"] = partial(
+                metrics.xgb_quantile_obj, quantile=quantile
+            )
+
+            # Initialize xgb model
+            specific_quantile_model = XGBRegressor()
+
+            # Set hyperparameters
+            specific_quantile_model.set_params(params=params_quantile)
+
+            # Train model for this specific quantile
+            quantile_models[quantile] = specific_quantile_model.fit(x, y)
+
+        # Update state of the estimator
+        self.estimators_ = quantile_models
+        self.is_fitted_ = True
+
+        return self
+
+    def predict(self, x: np.array, quantile: float = 0.5) -> np.array:
+        """Makes a prediction for a desired quantile
+
+        Args:
+            x np.array: Feature matrix
+            quantile float: Quantile for which a prediciton is desired,
+            note that only quantile are available for which a model is trained
+
+        Returns:
+            np.array with prediction
+
+        Raises:
+            ValueError in case no model is trained for the requested quantile
+
+        """
+        # Check if model is trained for this quantile
+        if quantile not in self.quantiles:
+            raise ValueError("No model trained for requested quantile!")
+
+        # Check/validate input
+        check_array(x)
```

Comment:
Dot these checks raise additional exceptions? If so it should be in the docstring ideally.

Reply:
Yes but I dont know which ones in which cases

Reply:
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html sklearn documentation is also not clear on which Exceptions are raised